contracts/contracts/harvest/Dripper.sol

```diff
@@ -0,0 +1,116 @@
+ * events don't have to call anything on this contract or pay any extra gas.
+ * Collect() is already be paying for a single write, since it has to reset
+ * the lastCollect time.
+ * - By having a collectAndRebase method, and having our external systems call
+ * that, the OUSD vault does not need any changes, not even to know the address
+ * of the dripper.
+ * - A rejected design was to retro-calculate the drip rate on each collect,
+ * based on the balance at the time of the collect. While this would have
+ * required less state, and would also have made the contract respond more quickly
+ * to new income, it would break the predictability that is this contract's entire
+ * purpose. If we did this, the amount of fundsAvailable() would make sharp increases
+ * when funds were deposited.
+ * - When the dripper recalculates the rate, it targets spending the balance over
+ * the duration. This means that every time that collect is is called, if no
+ * new funds have been deposited the duration is being pushed back and the
+ * rate decreases. This is expected, and ends up following a smoother but
+ * longer curve the more collect() is called without incoming yield.
+ *
+ */
+
+contract Dripper is Governable {
+    using SafeERC20 for IERC20;
+
+    struct Drip {
+        uint64 lastCollect; // overflows 262 billion years after the sun dies
+        uint192 perBlock; // drip rate per block
+    }
+
+    address immutable vault; // OUSD vault
+    address immutable token; // token to drip out
+    uint256 public dripDuration; // in seconds
+    Drip public drip; // active drip parameters
+
+    constructor(address _vault, address _token) {
+        vault = _vault;
+        token = _token;
+    }
+
+    function availableFunds() external view returns (uint256) {
+        uint256 balance = IERC20(token).balanceOf(address(this));
+        return _availableFunds(balance, drip);
+    }
+
+    function collect() external {
+        _collect();
+    }
+
+    function collectAndRebase() external {
+        _collect();
+        IVault(vault).rebase();
+    }
+
+    function setDripDuration(uint256 _durationSeconds) external onlyGovernor {
+        require(_durationSeconds > 0, "duration must be non-zero");
+        dripDuration = uint192(_durationSeconds);
+        Dripper(this).collect(); // duration change take immediate effect
+    }
+
+    function transferToken(address _asset, uint256 _amount)
+        external
+        onlyGovernor
+    {
+        IERC20(_asset).safeTransfer(governor(), _amount);
+    }
+
+    function _availableFunds(uint256 _balance, Drip memory _drip)
+        internal
+        view
+        returns (uint256)
+    {
+        uint256 elapsed = block.timestamp - _drip.lastCollect;
+        uint256 allowed = (elapsed * _drip.perBlock);
+        return (allowed > _balance) ? _balance : allowed;
+    }
+
+    function _collect() internal {
+        // Calculate send
```

Comment:
I am slightly worried an attacker (or competing product) can spend relatively small amount of gas (by frequently calling collect) to artificially keep our APY low. Stalling the funds in the dripper.

Seems like we can defend agains that by adding a check where collect fails if there is not a minimum difference between the last collect and the current one.

Reply:
Nice thanks a lot for supporting this with the charts! I was expecting a greater difference with calling it more frequently but like you point out (with the polygon analogy) it comes down more to approximation than to fundamentally changing the behaviour.